src/lib/Microsoft.Health.Events/EventConsumers/Service/EventBatchingService.cs

```diff
@@ -0,0 +1,139 @@
+        {
+            return _eventQueues.ContainsKey(queueId);
+        }
+
+        private EventQueue CreateQueueIfMissing(string queueId, DateTime initTime, TimeSpan flushTimespan)
+        {
+            return _eventQueues.GetOrAdd(queueId, new EventQueue(queueId, initTime, flushTimespan));
+        }
+
+        public Task ConsumeEvent(Event eventArg)
+        {
+            EnsureArg.IsNotNull(eventArg);
+
+            var queueId = eventArg.PartitionId;
+            var eventEnqueuedTime = eventArg.EnqueuedTime.UtcDateTime;
+
+            if (eventArg is MaximumWaitEvent)
+            {
+                if (EventQueueExists(queueId))
+                {
+                    var windowThresholdTime = GetQueue(queueId).GetQueueWindow();
+                    ThresholdWaitReached(queueId, windowThresholdTime);
+                }
+            }
+            else
+            {
+                var queue = CreateQueueIfMissing(queueId, eventEnqueuedTime, _flushTimespan);
+
+                queue.Enqueue(eventArg);
+
+                var windowThresholdTime = queue.GetQueueWindow();
+                if (eventEnqueuedTime > windowThresholdTime)
+                {
+                    ThresholdTimeReached(queueId, eventArg, windowThresholdTime);
+                    return Task.CompletedTask;
+                }
+
+                if (queue.GetQueueCount() >= _maxEvents)
+                {
+                    ThresholdCountReached(queueId);
+                }
+            }
+
+            return Task.CompletedTask;
+        }
+
+        // todo: fix -"Collection was modified; enumeration operation may not execute."
+        private async void ThresholdCountReached(string queueId)
+        {
+            Console.WriteLine($"The threshold count {_maxEvents} was reached.");
+            var events = await GetQueue(queueId).Flush(_maxEvents);
+            await _eventConsumerService.ConsumeEvents(events);
+            UpdateCheckpoint(events);
+        }
+
+        private async void ThresholdTimeReached(string queueId, Event eventArg, DateTime windowEnd)
+        {
+            Console.WriteLine($"The threshold time {_eventQueues[queueId].GetQueueWindow()} was reached.");
+            var queue = GetQueue(queueId);
+            var events = await queue.Flush(windowEnd);
+            await _eventConsumerService.ConsumeEvents(events);
+            queue.IncrementQueueWindow(eventArg.EnqueuedTime.UtcDateTime);
+            UpdateCheckpoint(events);
+        }
+
+        private async void ThresholdWaitReached(string queueId, DateTime windowEnd)
+        {
+            if (windowEnd < DateTime.UtcNow.AddSeconds(_timeBuffer))
+            {
+                Console.WriteLine($"Threshold wait reached. Flushing {_eventQueues[queueId].GetQueueCount()} events up to: {windowEnd}");
+                var events = await GetQueue(queueId).Flush(windowEnd);
+                await _eventConsumerService.ConsumeEvents(events);
```

Comment:
We need to increment the window time here as well correct?  If we don't any new event that arrives will immediately trigger a new batch.

Reply:
I don't think we need to increment here. This method is only called when we haven't received an event in a while, and will only flush if the window end is sufficiently in the past.

If we get a new event then it has to be outside the current window end, we will flush the previous window which has 0 events and advance the window to contain this new event.

The main reason why I didn't want to increment here is to avoid a race condition with the other flushing mechanisms (ThresholdTimeReached and ThresholdCountReached). Perhaps this should be better designed though

Reply:
Is there a race condition they should be processed on the same thread? We had a race condition in my version since we had a timer triggering the processing.